Daria

LDDmarc opened the pull request:
1) Данные
Реализовали модель данных и взаимосвязи с помощью CoreData.
Подписали все сущности под протокол Decodable для работы с JSON.
Затем поменяли подход чтения данных и данный протокол уже не понадобился. 
Подключили под SwiftyJSON.

Насторили получение (с сервера) и считывание данных и их обработку/обновление в бэкгрануде. Для этого написали собственный протокол и подписали каждую сущность CoreData под него. 

Оптимизировали запросы данных (уложили в один реквест)

Научились (ну хотя бы частично) пользоваться FetchResultsController. С помощью него настроили автоматическую анимацию таблиц (в случае если наши данные изменяются). 

Добавили активитииндикаторы и  ПулТуРефреш главным экранам. 

2) UI - проверили на iPhone SE. У всех экранов таб бара есть пулТуРефреш. Все экраны адекватно смотрятся в дарк мод (подводит только логотип реал мадрида)

2/1) Экран КОМАНДЫ

Список всех команд с поисковой строкой по имени команды. 

Detail экран команды: Добавили стэквью последних матчей, расположили весь ui так, чтобы он умещался на любом экране. А так же при скроле этого экрана, в правом верхнем углу появляется логотип команды

Так же на этом экране список матчей данной команды, отсортированный по дате

2/2) Экран МАТЧИ

Список матчей всх турниров с поисковой строкой (по имени команды) и segmentedControl. Матчи отсортированы по дате и сгруппированы по турнирам (визувлизируется с помощью section headers). Здесь так же был редизайн, и теперь названия команд всегда цмещаются в клетку на любом экране.

2/3) Экран ТУРНИРЫ

Картинка, даты, краткая информация о турнире и возможность просмотра соответствующих данному турниру команд, матчей и результатов. Клетка расширяется и собирается по тапу на нее.   

3/3) Экран РЕЗУЛЬТАТЫ

Результаты по турнирам в двух формах: таблица и форма, выбрать можно с помощью segmentedControl. Выбрать турнир можно с помощью RightTabBarItem. Так как матчей не может быть много, выбрали стандартный actionsheet.

LDDmarc committed:
Transition to remote server
- added NetworkManager, DataProvider, DataPresentation
- MatchesTableViewController embed in MatchesNavigationViewController
- deleted DataProcessing
- deleted all arrays in viewControllers

LDDmarc committed:
Animate changes in Core Data
- added generic functions for get data from json file
- added SwiftyJSON
- added update(with: JSON) for all entities, which team is contained
- ! added function beginUpdates and endUpdates for FetchResultsController (it is important for correct work of animating updates in tableView)

LDDmarc committed:
New kind of getting data
- other functions for url requests
- added tabbariitems
- added id and modified for all CoreData entities
- new json files
- code review

LDDmarc committed:
Added CoreData files

LDDmarc committed:
changing json

LDDmarc committed:
Redesign Teams and Matches screens
- optimized getting and updating data
- fixed bug of pull to refresh for teams

LDDmarc committed:
Redesign Tournament screen
- fixed bugs: labels colors in dark mode, incorrect accordance between teams and score
- added custom section header for ResultsTableViewController
- selected sizes for ResultsTableViewController
- all screens tested on iPhone SE

LDDmarc committed:
Design of Results screen
- added rightBarurronItem for choosing tournaments in ResultsTableViewController
- added custom Section Headers
- added teo kinds of result cells (table and form)

LDDmarc committed:
- Added "Pull to refresh for all tab bar screens"
- Added star buttons for matches

LDDmarc committed:
Refactoring

skipperprivate commented on LocalFootball/Team+CoreDataClass.swift:
```diff
@@ -73,7 +79,59 @@ public class Team: NSManagedObject, Decodable {
+           
+            let tournamentsStatisticsSet = NSMutableOrderedSet()
+            
+            for _ in 0..<self.teamTournamentsNames.count {
```
лучше вынести self.teamTournamentsNames.count в отдельную переменную чтобы в цикле каждый раз за ней не ходить

skipperprivate commented on LocalFootball/Teams/DetailTeamTableViewController.swift:
```diff
@@ -11,13 +11,27 @@ import CoreData
+            if let teamMatches = team.matches {
+                var sortedMatches = teamMatches.compactMap { $0 as? Match }
+                sortedMatches.sort(by: { (match1, match2) -> Bool in
+                    match1.date!.compare(match2.date!) == .orderedDescending
```
ну я бы безопасно анрапил

skipperprivate commented on LocalFootball/Teams/DetailTeamTableViewCell.swift:
```diff
@@ -10,26 +10,39 @@ import UIKit
+        selectionStyle = .none
+        
+        for matchView in [match1View, match2View, match3View, match4View, match5View] {
+            matchView?.layer.cornerRadius = 2 * .pi
```
2 * .pi тоже можно вычислить 1 раз вместо 5

skipperprivate commented on LocalFootball/Team+CoreDataClass.swift:
```diff
@@ -73,7 +79,59 @@ public class Team: NSManagedObject, Decodable {
+        if self.teamStatistics?.tournamentsStatistics?.count != self.teamTournamentsNames.count {
+            if let currentCount = self.teamStatistics?.tournamentsStatistics?.count {
+                for i in 0..<currentCount {
+                    context.delete((self.teamStatistics?.tournamentsStatistics![i])! as! NSManagedObject)
```
ну тоже - форс анрапов надо стараться избегать

skipperprivate commented on LocalFootball/Results/ResultsTableViewController.swift:
```diff
@@ -0,0 +1,275 @@
+    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        let headerView = tableView.dequeueReusableHeaderFooterView(withIdentifier: "ResultsTableSectionHeader") as! ResultsTableSectionHeader
+        headerView.tournamentNameLabel.text = currentTournamentName
+        headerView.contentView.backgroundColor = UIColor.systemGray5
```
раз уж наткнулся на цвета - вы приложение под темную тему адаптировать будете? если да - то цвета может придется подкорректировать, нет - тогда запретите темную тему. Не знаю будут ли это проверять, но я бы задумался

skipperprivate commented on LocalFootball/Matches/MatchesTableViewController.swift:
```diff
@@ -30,47 +62,82 @@ class MatchesTableViewController: UITableViewController {
+        }
+        dataProvider.fetchAllData { (error) in
+            guard error == nil else { return }
+            DispatchQueue.main.async {
```
а нормально отрабатывает main.async в main.asyncAfter? интересная "конструкция"

skipperprivate commented on LocalFootball/DataProcessing/NetworkManager.swift:
```diff
@@ -0,0 +1,58 @@
+class NetworkManager {
+    
+    static let shared = NetworkManager()
+    private init() {}
```
все в перемешку - переменные, методы, опять переменные. Я бы сгруппировал

LDDmarc committed:
LoadMore
- changed jsons
- added "load more" for matches in cellWillDisplay
- commented FetchResultsController  animation for matches because of table blincking
- removed sections in Matches screen.   (!!!) for furure: section name must be the first sortdescriptor

LDDmarc committed:
Relationships bug
 - fixed delete rule between team and its statistics. remove nsbatchdeleterequest for team

LDDmarc committed:
SDWebImage
- added pod SDWebImage
- removed imageData for Tournaments from CoreData
- set imageView for Tournament Cell by SDWebImage

skipperprivate commented on LocalFootball/Model/Team+CoreDataClass.swift:
```diff
@@ -127,7 +127,7 @@ public class Team: NSManagedObject, UpdatableManagedObject {
-            let tournamentsStatisticsSet = NSMutableOrderedSet()
+       //     let tournamentsStatisticsSet = NSMutableOrderedSet()
             
             for _ in 0..<teamTournamentsIds.count {
```
тоже лучше вынести в отдельную переменную
